cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -427,6 +453,28 @@ func (c *Client) AddMember(name string, peerAddrs string) ([]Member, error) {
 	return ret, nil
 }
 
+func memberPromoteWithRety(ctx context.Context, cli *clientv3.Client, learnerID uint64) error {
+	klog.V(1).Infof("[etcd] Promoting a learner as a voting member: %016x", learnerID)
+	// The interval and timeout are following how talos does. See:
+	// https://github.com/siderolabs/talos/blob/474604cd279def7a6798e24ede27feef955ba5a3/internal/app/machined/pkg/system/services/etcd.go#L598-L599
+	err := wait.PollImmediate(15*time.Second, 10*time.Minute, func() (bool, error) {
```

Comment:
i think we should use our own timeouts and remove thr comments about talos.
15 seconds / 10 minutes is a lot and doesn't align with the rest of kubeadm tomeouts?

what can we use here based on real experiment - i.e. how much time for promotion of 3 members? perhaps 1 seconds / 5 minutes or would that be too spammy?

Reply:
actually, there are two looping check here as I said here: https://github.com/kubernetes/kubernetes/pull/113318/files#r1049454549, do we really them all? based on what I observed the logic here seems problematic, if something wrong from `MemberPromote` it will return `err`, so this polling check `PollImmediate` would be redundant and won't take effective at all, or you can revert the change here to return `nil` and let it retry (we should consider the worse case, the long time to wait for a negative result).

I'd suggest to leave one polling check with the `period` and `timeout` properly defined.

Reply:
```
	// only retry for add member: MemberAddAsLearner or MemberAdd
	err = wait.ExponentialBackoff(etcdBackoff, func() (bool, error) {
		if isLearner {
			resp, err = cli.MemberAddAsLearner(ctx, []string{peerAddrs})
			if err == nil {
				err = memberPromoteWithRety(ctx, cli, learnerID)
				if err != nil {
					// return err as memberPromoteWithRety already did some retries.
					return false, err
				}
				return true, nil
			}
		}

		resp, err = cli.MemberAdd(ctx, []string{peerAddrs})
		if err == nil {
			return true, nil
		}
	}
```

I updated the PR to avoid dup waiting loop.
- ExponentialBackoff for MemberAddAsLearner or MemberAdd
- memberPromoteWithRety retries itself

Reply:
ok, i wonder if we can just add the add learner + promotion the main exp backoff loop and increase its end time / steps.

perhaps if the member is already added as learner (we need to check) then we can just retry the promotion.

nested loops are indeed problrmatic and don't make sense, since for example the outer loop is short and the inner loop for promotion is long.

Reply:
> since for example the outer loop is short and the inner loop for promotion is long.

outer loop is checked by steps instead of a time period, so it's not really a problem.

+1 to simplify the nested loops if it's not necessary.

Reply:
Got that. I removed the nested loops now.